### ISSUE TITLE
dotnet CLI: Allow cancellation for long-running commands

### ISSUE THREAD

**Issue by MiYanni** [dotnet/sdk#34416]
Related: https://github.com/dotnet/sdk/issues/33417
Related: https://github.com/dotnet/sdk/issues/31872

## Summary
Cancellation support for CLI commands is spotty and inconsistent currently (from what I'm told). Some preliminary research suggests that we can consistently handle cancellation by [weaving a `CancellationToken` throughout the system](https://learn.microsoft.com/en-us/dotnet/standard/commandline/handle-termination) and hooking that in System.CommandLine. We wouldn't need to do this for *every* command, but we should have it for commands that are long-running.

## High-Level Tasks
- [ ] Determine a comprehensive list of long-running commands
- [ ] Filter the list to those commands we'd want cancellation support for
- [ ] Add `CancellationToken` to base command classes in those commands we want to support
- [ ] Make sure support for cancellation extends to those actions external to the CLI (MSBuild/NuGet/MSI/etc)
- [ ] Determine a way to add tests for cancellation

***Feel free to add more tasks as necessary***

**Comment by MiYanni** [comment]
> > We wouldn't need to do this for every command, but we should have it for commands that are long-running.
> 
> doesn't hurt to have it for every command though. one man's short-running command is other man's long-running one

Agreed. But I'm trying to keep the scope *reasonable* for a first pass at doing this kind of work. So, targeting what is currently long-running would flush out any issues in the changes/design. If it works well, we could consider the work to build it into every command by default.

**Comment by OronDF343** [comment]
A good place to start would be `dotnet restore`, as that can take a very long time if not authenticated against the feed or some network issue exists. Waiting for the task to finish effectively means that Ctrl+C has no effect.

```
/usr/share/dotnet/sdk/6.0.415/NuGet.targets(132,5): warning MSB4220: Waiting for the currently executing task "RestoreTask" to cancel.
```
